streamed_identifiers.py

```diff
@@ -0,0 +1,99 @@
+"""
+A class for caching Streamed Identifiers. This class is designed to be inherited.
+
+Attributes:
+    identifiers_snapshot (Dict[str, Union[Asset, Conversation]]): A dictionary mapping IDs to their corresponding API call results.
+    identifiers_queue (queue.Queue): A queue for IDs to be processed.
+    identifiers_set (set): A set for IDs currently in the queue.
+    api_call (Callable[[str], Union[Asset, Conversation]]): A callable that takes an ID and returns either an Asset or a Conversation.
+    block (bool): A flag to indicate whether to wait for the queue to receive the first ID.
+    first_shot (bool): A flag to indicate if it's the first time to open the websocket.
+    lock (threading.Lock): A lock for thread safety.
+    worker_thread (threading.Thread): A thread for processing the queue.
+
+Methods:
+    worker(): Continuously processes IDs from the queue and updates the identifiers_snapshot.
+    update_identifier(id: str): Updates the identifier snapshot with the result of the API call.
+    streamed_identifiers_callback(ids: StreamedIdentifiers): Callback method to handle streamed identifiers.
+
+Example:
+    class AssetSnapshot(StreamedIdentifiersCache,api_call=AssetApi(PiecesSettings.api_client).asset_snapshot):
+        pass
+"""
+
+import queue
+import threading
+from typing import Dict, Union, Callable
+from pieces_os_client import Conversation, StreamedIdentifiers, Asset
+import sublime
+
+
+class StreamedIdentifiersCache:
+    """
+    This class is made for caching Streamed Identifiers.
+    Please use this class only as a parent class.
+    """
+
+    def __init_subclass__(cls, api_call: Callable[[str], Union[Asset, Conversation]], **kwargs):
+        super().__init_subclass__(**kwargs)
+        cls.identifiers_snapshot: Dict[str, Union[Asset, Conversation,None]] = {}  # Map id:return from the api_call
+        cls.identifiers_queue = queue.Queue()  # Queue for ids to be processed
+        cls.identifiers_set = set()  # Set for ids in the queue
+        cls.api_call = api_call
+        cls.block = True  # to wait for the queue to receive the first id
+        cls.first_shot = True  # First time to open the websocket or not
+
+
+    @classmethod
+    def sort_first_shot(cls):
```

Comment:
do we need this method?

Reply:
yupp we need it when we inherit sometimes I need to preform some sorts to sort the conversations for example but in the assets I don't need to sort them.

this class for inheritance only!